ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,478 @@
+    using ScaleCodec for uint16;
+
+    /* Events */
+
+    /**
+     * @dev Emitted when a pre-submission request is validated
+     * @param id the identifier for the submission request
+     * @param sender The address of the sender
+     */
+    event NewRequest(uint256 id, address sender);
+
+    /**
+     * @dev Emitted when the MMR root is updated
+     * @param mmrRoot the updated MMR root
+     * @param blockNumber the beefy block number of the updated MMR root
+     */
+    event NewMMRRoot(bytes32 mmrRoot, uint64 blockNumber);
+
+    /* Types */
+
+    /**
+     * @dev The Commitment, with its payload, is the core thing we are trying to verify with
+     * this contract. It contains a MMR root that commits to the polkadot history, including
+     * past blocks and parachain blocks and can be used to verify both polkadot and parachain blocks.
+     * @param payload the payload of the new commitment in beefy justifications (in
+     * our case, this is a new MMR root for all past polkadot blocks)
+     * @param blockNumber block number for the given commitment
+     * @param validatorSetID validator set id that signed the given commitment
+     */
+    struct Commitment {
+        uint32 blockNumber;
+        uint64 validatorSetID;
+        Payload payload;
+    }
+
+    struct Payload {
+        bytes32 mmrRootHash;
+        bytes prefix;
+        bytes suffix;
+    }
+
+    /**
+     * @dev The ValidatorMultiProof is a collection of proofs used to verify a commitment signature
+     * @param signatures an array of validator signatures
+     * @param indices an array of the leaf indices
+     * @param addrs an array of each validator address
+     * @param merkleProofs an array of merkle proofs from the chosen validators
+     */
+    struct ValidatorMultiProof {
+        bytes[] signatures;
+        uint256[] indices;
+        address[] addrs;
+        bytes32[][] merkleProofs;
+    }
+
+    /**
+     * @dev The ValidatorProof is a proof used to verify a commitment signature
+     * @param signature validator signature
+     * @param index index of the validator address
+     * @param addr validator address
+     * @param merkleProof merkle proof for the validator
+     */
+    struct ValidatorProof {
+        bytes signature;
+        uint256 index;
+        address addr;
+        bytes32[] merkleProof;
+    }
+
+    /**
+     * @dev A request is used to link initial and final submission of a commitment
+     * @param sender the sender of the initial transaction
+     * @param commitmentHash the hash of the commitment they are claiming has been signed
+     * @param bitfield a bitfield signalling which validators they claim have signed
+     * @param blockNumber the block number for this commitment
+     */
+    struct Request {
+        address sender;
+        uint64 validatorSetID;
+        bytes32 commitmentHash;
+        uint256[] bitfield;
+        uint256 blockNumber;
+        ValidatorSet vset;
```

Comment:
looks unused?

Reply:
right but this is only used offchain, so adds extra gas costs in production with no real on-chain use. maybe createFinalBitfield can take the vset or a boolean as an input instead of it being in Request?

Reply:
Added https://linear.app/snowfork/issue/SNO-267/optimize-gas-cost-for-submitinitial to track gas optimizations.